tools/run_tests/run_xds_tests.py

```diff
@@ -557,114 +802,226 @@ def start_xds_client(service_port):
+        create_health_check_firewall_rule(gcp, firewall_name)
+        backend_service = add_backend_service(gcp, backend_service_name)
+        alternate_backend_service = add_backend_service(
+            gcp, alternate_backend_service_name)
+        create_url_map(gcp, url_map_name, backend_service, service_host_name)
+        create_target_http_proxy(gcp, target_http_proxy_name)
         potential_service_ports = list(args.service_port_range)
         random.shuffle(potential_service_ports)
         for port in potential_service_ports:
             try:
-                create_global_forwarding_rule(
-                    compute,
-                    PROJECT_ID,
-                    FORWARDING_RULE_NAME,
-                    port,
-                    target_http_proxy_url,
-                )
-                service_port = port
+                create_global_forwarding_rule(gcp, forwarding_rule_name, port)
+                gcp.service_port = port
                 break
             except googleapiclient.errors.HttpError as http_error:
                 logger.warning(
-                    'Got error %s when attempting to create forwarding rule to port %d. Retrying with another port.'
-                    % (http_error, port))
-        if not service_port:
+                    'Got error %s when attempting to create forwarding rule to '
+                    'port %d. Retrying with another port.' % (http_error, port))
+        if not gcp.service_port:
             raise Exception('Failed to pick a service port in the range %s' %
                             args.service_port_range)
-        template_url = create_instance_template(compute, PROJECT_ID,
-                                                TEMPLATE_NAME, service_port)
-        instance_group_url = create_instance_group(compute, PROJECT_ID, ZONE,
-                                                   INSTANCE_GROUP_NAME,
-                                                   INSTANCE_GROUP_SIZE,
-                                                   service_port, template_url)
-        add_instances_to_backend(compute, PROJECT_ID, BACKEND_SERVICE_NAME,
-                                 instance_group_url)
+        create_instance_template(gcp, template_name, args.network,
+                                 args.source_image)
+        instance_group = add_instance_group(gcp, args.zone, instance_group_name,
+                                            _INSTANCE_GROUP_SIZE)
+        patch_backend_instances(gcp, backend_service, [instance_group])
+        same_zone_instance_group = add_instance_group(
+            gcp, args.zone, same_zone_instance_group_name, _INSTANCE_GROUP_SIZE)
+        secondary_zone_instance_group = add_instance_group(
+            gcp, args.secondary_zone, secondary_zone_instance_group_name,
+            _INSTANCE_GROUP_SIZE)
     except googleapiclient.errors.HttpError as http_error:
-        if TOLERATE_GCP_ERRORS:
+        if args.tolerate_gcp_errors:
```

Comment:
Is there a better check that we can do with the error? (String matching?)

We don't want to tolerate _any_ http error, we only want to continue if the error is `"failed to create resource because it already exists"`.

Reply:
The HttpError class doesn't provide any details other than the string, and I'd prefer not to do string matching...this whole error recovery is quite brittle as-is, but otherwise it is incredibly slow to manually run/debug this script as you'd have to recreate the entire GCP resources each local run. I'm comfortable leaving this as-is for now as it's only for manual runs.